Re-ingest: add full option

sevein opened the pull request:
Redmine ticket: https://projects.artefactual.com/issues/9211
See also: artefactual/archivematica#384

I got some initial feedback as part of #86. I filed a new PR because the ticket and the branch of the name was wrong due to some misunderstanding on my side.

I think that what's left is to see if I got right the exception handling in `models/pipeline.py`. I tried to create a couple of meaningful exceptions but now I kind of feel that it's overcomplicated. Let me know what you think please.

sevein committed:
Re-ingest: add full option

mistydemeo commented on storage_service/locations/models/pipeline.py:
```diff
@@ -126,3 +128,76 @@ def create_default_locations(self, shared_path=None):
+            'username': self.api_username,
+            'api_key': self.api_key
+        }
+        if 'fields' in kwargs:
```
If this is the only thing you're using `kwargs` for, maybe make it a defined parameter with a default of `None`?

``` python
def _request_api(self, method, path, fields=None)
# ...

if fields is not None:
    fields.update(kwargs['fields'])
```

mistydemeo commented on storage_service/locations/models/pipeline.py:
```diff
@@ -126,3 +128,76 @@ def create_default_locations(self, shared_path=None):
+            resp = requests.request(method, url, data=fields, allow_redirects=True)
+        except requests.exceptions.RequestException, e:
+            LOGGER.error('Unable to connect to pipeline %s.', self)
+            et, ei, tb = sys.exc_info()
```
If you're not planning to use `et` or `ei`, mark that by making them `_`:

``` python
_, _, tb = sys.exc_info()
```

sevein commented:
Thank you! It's done now.

mistydemeo commented:
The exception handling looks fine to me.

sevein commented:
Should I merge?

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -894,6 +900,27 @@ def start_reingest(self, pipeline, reingest_type):
+                config = pipeline.get_processing_config(processing_config)
+            except PipelineClientException:
+                LOGGER.exception('Reingest: processing configuration %s could not be loaded, error: %s', processing_config, e)
+                pass
```
Nitpick: pass not needed - it's a noop useful for marking an empty block, and the log line above means the block isn't empty.

sevein replied:
ok

Hwesta commented on storage_service/locations/models/package.py:
```diff
@@ -894,6 +900,27 @@ def start_reingest(self, pipeline, reingest_type):
+                    LOGGER.debug('Reingest: processing configuration %s written, location: %s', processing_config, config_path)
+                    reingest_files.append(os.path.join(relative_path, 'processingMCP.xml'))
+                except IOError as e:
+                    LOGGER.exception('Reingest: processing configuration %s could not be written, error: %s', processing_config, e)
```
LOGGER.exception will automatically log the exception details including name, so it doesn't need to be written in the error message.

If this error happens, do we want to continue processing, or should this re-raise?

sevein replied:
Good points. I'll re-raise the exception.

Hwesta commented on storage_service/locations/models/pipeline.py:
```diff
@@ -126,3 +128,76 @@ def create_default_locations(self, shared_path=None):
+        if resp.status_code != requests.codes.ok:
+            raise PipelineClientUnexpectedCodeException(self, resp.status_code)
+        if not resp.text:
+            raise PipelineClientException('Pipeline {}: empty processing configuration ({}).'.format(self, name))
```
This or the docstring should be updated to reflect the actual empty config behaviour.

Hwesta commented:
> I think that what's left is to see if I got right the exception handling in models/pipeline.py. I tried to create a couple of meaningful exceptions but now I kind of feel that it's overcomplicated. Let me know what you think please.

While not wrong, the exception handling looks odd to me. I'd prefer to see the requests.Exception in a traceback, since that's clearer to me what went wrong.  I also wouldn't check for non-Falsey parameters, since subsequent calls that use them will fail.  With such a small use case I would not create custom exceptions, especially since not doing it simplifies the code.  I like the `_request_api` cleanup.

sevein commented:
Cool, I followed your recommendations. Let me know what you think.

Hwesta committed:
Fix typo

Hwesta committed:
Reingest: Update preservation derivatives

Add new preservation derivatives and delete old ones to the stored AIP.
Update the bag manifest for the new & removed files.

Hwesta committed:
Reingest: Return new unit's uuid

Hwesta committed:
Reingest: Only store first and last METS

When storing a reingested AIP, keep the backup METS if there isn't one
already.  This should keep the original AIP METS, and the most recently
updated METS.

Hwesta committed:
Space: os.rename sets permissions

move_rsync can optimize the move by using os.rename. However, rsync also sets
permissions when it copies. Add permission setting to the os.rename paths.

Hwesta committed:
Reingest: Fix issues when files are moved instead of copied

Handle when moved files don't leave copies behind, eg when move_rsync is
called with try_mv_local=True

jrwdunham commented on storage_service/locations/forms.py:
```diff
@@ -219,16 +219,22 @@ def __init__(self, *args, **kwargs):
     reingest_type = forms.ChoiceField(choices=REINGEST_CHOICES, widget=forms.RadioSelect)
+    processing_config = forms.CharField(required=False, initial='default',
+        help_text='Only needed in full re-ingest',
+        widget=forms.TextInput(attrs={'placeholder': 'default'}))
```
Is the help text "Only needed in full re-ingest" accurate? I believe the behaviour is that the processing config will only work in full reingest and will not work at all in the other types of ingest. Is this not true? If it is true, then maybe the help_text should be something like "Optional; the processing config will only work with full re-ingests."

Hwesta replied:
Updated text

Hwesta commented:
Looks good to me, but someone should review the changes I made.

jrwdunham commented on storage_service/locations/models/package.py:
```diff
@@ -941,12 +946,32 @@ def start_reingest(self, pipeline, reingest_type):
+        # include the file in reingest_files.
+        if reingest_type == self.FULL and processing_config != 'default':
+            try:
+                config = pipeline.get_processing_config(processing_config)
```
I see that the dashboard API does not expose a GET processing-configurations/ endpoint. However, if it did, we could call it before the re-ingest form was generated and then only allow the user to select from available configs. then this get_processing_config call would rarely (never?) fail. Wishful thinking, but thought I'd mention it.

Hwesta replied:
I like that idea!  I think it's out of scope for right now though. The processing-config endpoint is added in artefactual/archivematica#313 so should be exposed in qa/1.x's dashboard API.

jrwdunham commented on storage_service/locations/models/package.py:
```diff
@@ -1112,9 +1141,34 @@ def finish_reingest(self, origin_location, origin_path, reingest_location, reing
+                    # Copy new preservation derivative
+                    LOGGER.info('Moving %s to %s', reingest_preservation_path, original_preservation_path)
+                    shutil.copy2(reingest_preservation_path, original_preservation_path)
+
```
While it's an edge case, I think this block will cause a bug if a transfer contains a file whose original name has a UUID suffix. For example, try to re-ingest a SIP whose original transfer contains a file named "Nemastylis_geminiflora_Flower-a464e051-b6db-46cf-9f45-64c00ae16171.PNG". I tried this and the resulting AIP (after re-ingest) did not contain the .tif derivative that was created during the original ingest.

Assuming I'm correct in identifying this as an error, maybe a better way of identifying preservation derivatives would be to look for files whose names are suffixed by UUIDs AND where there is another file with the same basename, minus the UUID suffix.

Hwesta replied:
Nice catch.  Updated so the regex matches on the extension as well (so originalfile-originaluuid.ext doesn't match both originalfile-originaluuid.ext and originalfile-originaluuid-preservationuuid.ext), and added a check to not delete itself if detected as a duplicate preservation file.  I'll rebase the fix into the original commit if it looks good.

jrwdunham reviewed (commented):
Two minor comments and one possible edge-case bug.

jrwdunham commented:
👍
